File: Data Structures & Algorithms/minimum-stack/submission-2.go
type MinStack struct {
    stk []int
    minHistory []int
}


func Constructor() MinStack {
    return MinStack{stk : []int{}, minHistory: []int{}}
}


func (this *MinStack) Push(value int)  {
    this.stk = append(this.stk, value)
    if len(this.minHistory) == 0 {
        this.minHistory = append(this.minHistory, value)
    } else {
        n := len(this.minHistory)
        this.minHistory = append(this.minHistory, min(this.minHistory[n-1], value))
    }
}


func (this *MinStack) Pop()  {
    n := len(this.stk)
    this.stk = this.stk[:n-1]
    n = len(this.minHistory)
    this.minHistory = this.minHistory[:n-1]
}


func (this *MinStack) Top() int {
    n := len(this.stk)
    return this.stk[n-1]
}


func (this *MinStack) GetMin() int {
    n := len(this.minHistory)
    return this.minHistory[n-1]
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(value);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: Data Structures & Algorithms/online-stock-span/submission-2.py
class StockSpanner:

    def __init__(self):
        self.stack = []
        self.i = 0 # index of next day

    def next(self, price: int) -> int:
        while self.stack and price >= self.stack[-1][0]:
            self.stack.pop()
        
        if self.stack:
            _, j = self.stack[-1]
            curSpan = self.i - j
        else:
            curSpan = self.i + 1
        
        self.stack.append((price, self.i))
        self.i += 1
        return curSpan
        


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)